src/server.c

```diff
@@ -572,27 +572,21 @@ static void monitor_cb(uv_prepare_t *monitor)
 {
 	struct dqlite_node *d = monitor->data;
 	int state = raft_state(&d->raft);
-	/*
 	queue *head;
 	struct conn *conn;
-	*/
 
 	if (state == RAFT_UNAVAILABLE) {
 		return;
 	}
 
-	/* TODO: we should shutdown clients that are performing SQL requests,
-	 * but not the ones which are doing management-requests, such as
-	 * transfer leadership.  */
-	/*
 	if (d->raft_state == RAFT_LEADER && state != RAFT_LEADER) {
+		tracef("node %llu@%s: leadership lost", d->raft.id, d->raft.address);
 		QUEUE__FOREACH(head, &d->conns)
 		{
 			conn = QUEUE__DATA(head, struct conn, queue);
-			conn__stop(conn);
+			gateway__leader_close(&conn->gateway, RAFT_LEADERSHIPLOST);
```

Comment:
The `RAFT_LEADERSHIPLOST` error should actually be generated only by the raft library.

To put it shortly, by the time `monitor_cb` gets called, the raft library should have already triggered a cascade of callbacks that terminates any pending dqlite client request with `LEADERSHIP_LOST`. So a good part of the cleanups happening in `gateway__close` should not be needed, and it would be a bug if they are.

Basically the detailed sequence of events should be:

1. The raft library decides that the node needs to step down from leader state. As part of converting from leader to follower it will terminate all pending apply requests with `RAFT_LEADERSHIPLOST` (see `convertClearLeader()` in `convert.c`).
2. The `leaderApplyFramesCb()` callback in dqlite's `leader.c` gets called for inflight commit requests, which will set `g->leader->inflight` to `NULL` and call `leader_exec_cb()` in `gateway.c` via `maybeExecDone()` .
3. The `leader_exec_cb()` sets `g->req` and `g->stmt` to NULL
4. At the point there should be no state left regarding inflight write client requests, however I guess there might be an pending query statement if it's being batched (see below).
5. The `monitor_cb` gets called after all this has already happened and should not deal with terminating write requests.

So one thing that most probably needs to be fixed is that `gateway__leader_close()` should assert that there are no inflight write requests (and any associated state).

However, one issue of the overall approach that maybe we haven't considered (or at least I didn't) is what happens when the leadership is lost in the middle of batched query response. In that case I believe the server will never send to the client the next batch and the client will be stuck waiting for it. There should be a timeout for that, but it feels a bit sub-optimal.

Probably we need to reason a bit about this case.

Reply:
Good to know that those barrier cleanups should not be needed, they are triggered by tests though so should be easy enough to fix in raft (I hope) when they are removed.

Thanks for laying out what the planned sequence of events is/was, makes things clear.

I will let this one rest then for a while though and see when I can continue work on it, because I need to work on other things related to the cycle.

Reply:
Sounds good, there are a number of details involved, more than we envisioned. I was also thinking about the implications for a future proper handling of `LEADERSHIP_LOST` via client sessions and request IDs, and I'm not totally sure about how the whole mechanics work with respect to this issue.

Just to be clear, those cleanups **are** needed when `gateway__close()` gets called "normally", they are not needed only when called in this case. However, I'll add that those cleanups are still basically hacks and the proper fix would be to make `gateway__close()` asynchronous, so all callbacks have a chance to get triggered instead of having them triggered "by hand" (hope it's clear what I mean). There's a comment about that in [`leaderApplyFramesCb`](https://github.com/canonical/dqlite/blob/master/src/leader.c#L254):

Reply:
> The `RAFT_LEADERSHIPLOST` error should actually be generated only by the raft library.
> 
> To put it shortly, by the time `monitor_cb` gets called, the raft library should have already triggered a cascade of callbacks that terminates any pending dqlite client write request with `LEADERSHIP_LOST`. So a good part of the cleanups happening in `gateway__close` should not be needed, and it would be a bug if they are.
> 
> Basically the detailed sequence of events should be:
> 
>     1. The raft library decides that the node needs to step down from leader state. As part of converting from leader to follower it will terminate all pending apply requests with `RAFT_LEADERSHIPLOST` (see `convertClearLeader()` in `convert.c`).
> 
>     2. The `leaderApplyFramesCb()` callback in dqlite's `leader.c` gets called for inflight commit requests, which will set `g->leader->inflight` to `NULL` and call `leader_exec_cb()` in `gateway.c` via `maybeExecDone()` .
> 
>     3. The `leader_exec_cb()` sets `g->req` and `g->stmt` to NULL
> 
>     4. At the point there should be no state left regarding inflight write client requests, however I guess there might be an pending query statement if it's being batched (see below).
> 
>     5. The `monitor_cb` gets called after all this has already happened and should not deal with terminating write requests.
> 
> 
> So one thing that most probably needs to be fixed is that `gateway__leader_close()` should assert that there are no inflight write requests (and any associated state).
> 
> However, one issue of the overall approach that maybe we haven't considered (or at least I didn't) is what happens when the leadership is lost in the middle of batched query response. In that case I believe the server will never send to the client the next batch and the client will be stuck waiting for it. There should be a timeout for that, but it feels a bit sub-optimal.
> 
> Probably we need to reason a bit about this case.

I would be inclined to merge it anyhow as it fixes https://github.com/canonical/dqlite/issues/355 that pops up quite regularly in the LXD testsuite. Unless you think it makes the current implementation worse.

Reply:
I don't think it makes things worse indeed, so I'm ok with that. We can revisit this later on.